keps/sig-apps/2307-job-tracking-wihout-lingering-pods/README.md

```diff
@@ -0,0 +1,531 @@
+  - [Scalability](#scalability)
+  - [Troubleshooting](#troubleshooting)
+- [Implementation History](#implementation-history)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [x] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [x] (R) KEP approvers have approved the KEP status as `implementable`
+- [x] (R) Design details are appropriately documented
+- [x] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [x] (R) Graduation criteria is in place
+- [x] (R) Production readiness review completed
+- [x] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+The current Job controller currently relies on completed Pods to not be removed
+in order to track the Job completion status. This proposal presents an
+alternative implementation for Job tracking that does not have this dependency.
+
+## Motivation
+
+The current approach of relying on the Pods existence is problematic for Jobs
+that require a big number of completions or for clusters with too many Jobs
+running at the same time. The finished Pods cannot be removed until the entire
+Job completes, even if the Pods failed.
+
+Furthermore, once the number of finished Pods reaches a threshold, the Pod
+garbage collection controller starts removing Pods. Today, the Job controller
+relies on the garbage collector having a big threshold.
+
+### Goals
+
+- Perform Job completion and failures tracking without relying on lingering
+  Pods.
+
+### Non-Goals
+
+- Remove Pods once they have been accounted for.
+
+## Proposal
+
+The Job controller creates Pods with a finalizer to prevent finished Pods from
+being removed by the garbage collector. The Job controller removes the finalizer
+from the finished Pods once it has accounted for them. In subsequent Job syncs,
+the controller ignores finished Pods that don't have the finalizer.
+
+### Notes/Constraints/Caveats (Optional)
+
+Due to the lack of support of atomic changes across kubernetes objects, an
+intermediate state is necessary. Before removing the Pod finalizers,
+the controller adds finished Pods to a list in the Job status. After removing
+the Pod finalizers, the controller clears the list and updates the counters.
+
+### Risks and Mitigations
```

Comment:
is increasing object size a risk that we should mention?

Reply:
yes, added.

Reply:
FYI, the KEP already mentions limiting the UIDs in the controller

Reply:
Thanks I see that now. Do we have estimates on how the caps interact with the maximum anticipated scale, and on how we can shard that work across multiple workers if necessary? That is, I still would like to understand:

1) the anticipated scale
2)the plan for handling the space and time complexity at that maximum scale

If we chunk into small units, that means it may take longer to process, etc.

Reply:
Let's say we remove 500 finalizers and do the corresponding status updates every 15s (this roughly corresponds with the default 50 QPS limit on the controller's client). That means that in 1 min, we can process 2000 terminated Pods. Does this sound reasonable?

Note that each finalizer removal happens in its own goroutine.

Reply:
I am not sure that will be fast enough for the most demanding of batch users.   Do you have a plan to get feedback on that?

From an API Server standpoint, it would be better to batch up hundreds of finalizer removals into a single update request.  Is it a lot of work for those goroutines to not do the update themselves but to instead send a message through a channel to something else that can batch up the updates?

Reply:
does the apiserver currently support batch updates?

Reply:
No, you have to do the finalizer removals individually since they are on different objects. But you should at least be able to use patch requests which won't have conflicts. And you can perform the requests in parallel and trust APF to keep you from overloading apiserver.

Reply:
Which are you doing:

Option 1:
- Say 100 pods recently finished.
- make 100 calls to apiserver for each pods resource.
- make 100 calls to apiserver for the job resource

Option 2:
- Say 100 pods recently finished.
- make 100 calls to apiserver for each pods resource.
- make 1 call to apiserver for the job resource

Reply:
option 2, but it's actually 2 api calls to the job resource

Reply:
Added expected number of processed Pods per minute and a target of 5000 for beta graduation.

Reply:
Thanks, and I see GA grad criteria has 10^5 for completions (for a single job I presume). So the job tracking takes about 20 minutes for a job of that size, in this plan, in the degenerate scenario where every pod just exits. Is that 5000 per job, or across all jobs? If there are 10 jobs of 100k pods running, accounting takes 200 minutes? Is that OK for our users? (really asking, I have no idea. it sounds reasonable to me).

Reply:
Clarified 5000 to be across any number of Jobs.

As for 10^5 criteria, clarified to be per Job and in the order of minutes. I think we can aim to be 10min, but we can tune that after alpha is implemented.